Add slow path for device

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77972
* #77971
* #77970
* #77969
* __->__ #77684

Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

Differential Revision: [D36618465](https://our.internmc.facebook.com/intern/diff/D36618465)

Commit:
Add slow path for device

[ghstack-poisoned]

Commit:
Update on "Add slow path for device"




[ghstack-poisoned]

Commit:
Update on "Add slow path for device"


Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -371,6 +371,22 @@ IntArrayRef TensorImpl::sizes_custom() const {
+    TORCH_CHECK(
+        false,
+        "cannot access PyObject for Tensor on interpreter ",
+        pyobj_interpreter_.load()->name());
```
The interpreter access and fail if null stuff should get factored into a method

Commit:
Update on "Add slow path for device"


Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

[ghstack-poisoned]

Commit:
Update on "Add slow path for device"


Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

[ghstack-poisoned]

Commit:
Update on "Add slow path for device"


Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618465).

Commit:
Update on "Add slow path for device"


Similar to https://github.com/pytorch/pytorch/pull/77396, add a way to override `device` on Tensor Subclasses by adding a bit to call into a custom, virtual `device` function. This is necessary to allow extending devices with a `meta` device to also carry along a device so as to compute device propagation in addition to other properties (FakeTensors).

Differential Revision: [D36618465](https://our.internmc.facebook.com/intern/diff/D36618465)

[ghstack-poisoned]
